Feature/smart computer

serpient opened the pull request:
# Smart Computer feature
## Smart Computer's Move Strategy:
1. Takes winning move if possible
2. Takes blocking move if opponent is close to win
3. center if available
4. corners if available
5. completes a row (a possible winning set missing 2 possible moves) if it had an existing token
6. Blocks a row if opponent has at least 2 matching tokens
6. then random if nothing else

```
    def move(board:, presenter:)
        @board = board
        @transformed_boards = BoardTransformer.transform_with_idx(board)
        @winning_token_count = board.size 

        take_winning || 
        take_blocking(row_token_count: winning_token_count - 1) || 
        take_center || 
        take_empty_corner || 
        take_row_completing_move(row_token_count: winning_token_count - 2) || 
        take_blocking(row_token_count: winning_token_count - 2) || 
        take_random_position
    end
```

# Game Play GIF
![3x3 board](https://media.giphy.com/media/U52AMpAcKDyfUhhgkM/giphy.gif)
![4x4 board](https://media.giphy.com/media/hQvG97LOaq6MUCkaQR/giphy.gif)

serpient committed:
smart computer feature

- extend diagonal transformer to properly work with 4x4 board
- create new player type
- smart computer has a strategy

serpient committed:
store player types as array

serpient committed:
refactor smart_computer for clarity

- prepend module name to method
- add new strategy for next optimal move -- computer tries to complete a row it had started.
- move board transformer array to board transformer module
- extract functions for clarity / naming / logic flow

serpient committed:
remove unused fn

serpient committed:
rename functions for clarity

serpient committed:
add additional blocking check in smart computer

serpient committed:
refine naming for functions and params

ssunday commented on lib/board.rb:
```diff
@@ -32,4 +28,20 @@ def empty_positions
         positions[position] == Token::EMPTY
     end
+
+    def positions_with_idx
```
`idx` -> `index`

ssunday commented on lib/board.rb:
```diff
@@ -32,4 +28,20 @@ def empty_positions
+    end
+
+    def center
+        size == 3 ? 4 : nil
```
`4 if size == 3` ? maybe?

ssunday commented on lib/board.rb:
```diff
@@ -32,4 +28,20 @@ def empty_positions
+        if size == 3
+            return [0, 2, 6, 8]
+        else
+            return [0, 3, 12, 15]
```
Don't need `return`...and can you make this so it's not hard coded by the size of the board?

ssunday commented on lib/player/smart_computer.rb:
```diff
@@ -0,0 +1,87 @@
+
+    private
+    attr_accessor :transformed_boards, :board
+    def take_winning
```
new line above

ssunday commented on lib/player/smart_computer.rb:
```diff
@@ -0,0 +1,87 @@
+    end
+
+    def find_row_with_token_count(token_to_find:, target_token_count:)
+        return transformed_boards.find do |transformed_board|
```
Don't need return

ssunday commented on lib/player/smart_computer.rb:
```diff
@@ -0,0 +1,87 @@
+            row_token_count[Token::EMPTY] += 1 if position.first == Token::EMPTY
+            row_token_count[Token::X] += 1 if position.first == Token::X
+        end
+        return row_token_count
```
return not needed

ssunday commented on spec/player_spec.rb:
```diff
@@ -44,4 +53,245 @@
+
+    context 'smart computer - winning move - 3x3' do
+        it 'takes diagonal winning move if possible' do
+            computer_player = Player.new(
```
you can put `computer_player` into a `let` since it's the same for each spec

ssunday commented on spec/player_spec.rb:
```diff
@@ -44,4 +53,245 @@
+            board.update(position: nil, token: nil, all_positions: smart_computer_board)
+            expect(computer_player.move(board: board, presenter: presenter)).to eq(1).or eq(2)
+        end
+        end
```
spacing

serpient committed:
refactor per feedback

- rename idx to index
- remove ternary operations if false returns nil. Use format like "4 if size == 3" instead of "size == 3 ? 4 : nil"
- remove unnecessary usage of returns (ruby automatically returns last statement)
- clean up test case setup in player spec. It's the same setup for most of the test cases, so moved into a before block
-

serpient committed:
adjust board corner generator + test cases

ssunday commented on spec/board_spec.rb:
```diff
@@ -74,11 +74,28 @@
         end
     end
+
+    context 'corners' do
```
👏

christophermanahan commented on lib/board.rb:
```diff
@@ -18,18 +18,30 @@ def full?
     def empty_positions
-        positions.each_with_index.reduce([]) do |empty_positions, (_position, idx)| 
-            empty_positions.push(idx) if position_empty?(idx)
+        positions.each_with_index.reduce([]) do |empty_positions, (_position, index)| 
```
Could this be simpler with map?

serpient replied:
I had thought a `map` would iterate over each item and transform it in some way. In this case, I'm looking through and picking out matching items so it felt like `each` or even `find` would be more appropriate?

ssunday replied:
Wouldn't select be better, then?

serpient replied:
I thought of using select, but I need to know each index in the array, and after filtering the values, only return the matching index. Which is why i did the `each_with_index` and then `reduce` to control what I wanted the function to output. Is that ok? Or find a way to use select instead ?

christophermanahan replied:
Lol look at me suggesting map because I am cutting back on coffee.
But `positions.each_index.select` could be pretty useful!

esmevane replied:
I third the `select` suggestion, but I wanted to chime in and add this:  `reduce` is, in a lot of these functional operations, getting used one way or the other.  So while this could get a little more readability by preferring `select`, there's a significant gain in Ruby-land by leveraging your reducers.

Random example:  If later you wind up needing to `select` and then `reduce` over that, it's better to just switch back over to a `reduce` entirely, because it can do both and that would halve your potential iterations.  Which could be a surprisingly useful optimization depending on how you choose to simulate things in your minimax.

serpient replied:
Just to make sure I'm doing this right -- when I use select instead of reduce, i can't control what i'm returning. AKA, it returns the entire unit of position/index instead of just the index.
Was there another way to use select to return the desired value?

So this function
```
   # positions is an array
    def empty_positions
        positions.each_with_index.select() do |_position, index| 
            index if position_empty?(index)
        end
    end
```

returns 
<img width="528" alt="Screen Shot 2019-05-06 at 7 13 13 AM" src="https://user-images.githubusercontent.com/29721784/57230979-7053d880-6fce-11e9-884f-d9c6bf67d78e.png">

esmevane replied:
Ah!  Okay, select alone probably doesn't work.  If you're filtering _and_ mapping then `reduce` is totally your huckleberry.

serpient committed:
clean up diagonal transformation

- different conditionals weren't actually needed. After adjusting the default generation of diagonal values, it actually takes care of all cases. No special conditions needed

christophermanahan commented on lib/player/player_types.rb:
```diff
@@ -1,4 +1,20 @@
+    COMPUTER = "D"
+    SMART_COMPUTER = "S"
+
+    def get_player(type)
```
A switch might be nice here. Up to you!

christophermanahan commented on lib/player/smart_computer.rb:
```diff
@@ -0,0 +1,87 @@
+        @transformed_boards = BoardTransformer.transform_with_index(board)
+        @winning_token_count = board.size 
+
+        take_winning || 
```
This is cool, I like the or operator for precedence.

christophermanahan commented on lib/player/smart_computer.rb:
```diff
@@ -0,0 +1,87 @@
+    attr_accessor :transformed_boards, :board
+
+    def take_winning
+        find_optimal_move(token_to_find: token, target_token_count: winning_token_count - 1)
```
This target_token_count is a bit confusing. It may be simpler to just try all available moves and see if any cause a win. For block you can try all available moves with the opponent's token and see if any cause a win. This count is an interesting idea though!

serpient replied:
This part has been really hard to name , which certainly doesn't help the confusion.
Essentially, if its blocking, for example:
- first it will block moves that is 1 away from a winning token count (for example, needs 4 adjacent tokens to win)
- then when it runs again on line 19, it will try to find a blocking move within a row that is 2 away from a winning token count.

same algorithm for the winning / row_completing moves, but I had wrapped it in named functions in hopes of clarifying the intention.

christophermanahan replied:
Maybe `tokens_until_win`?

christophermanahan reviewed (commented):
Lot's of cool/unique implementations in here! :)

serpient committed:
small refactor

- clean up unused code like computer not needed to output messages anymore
- move non public attributes to the private section
- trim message strings

serpient committed:
use switch statement in player type for readability

ssunday commented on lib/player/player_types.rb:
```diff
@@ -9,9 +9,16 @@ module PlayerType
-        return SmartComputer.new if type == SMART_COMPUTER
+        case type
+        when HUMAN
+            return Human.new
```
Don't need returns

serpient committed:
remove unnecessary return

esmevane commented on lib/board.rb:
```diff
@@ -18,18 +18,30 @@ def full?
     end
+
+    def positions_with_index
+        positions.map.with_index { |val, index| [val, index] }
```
Have you played around with structs much?  In a lot of languages, you can get some mileage out of fixed tuples, but in Ruby / dynamic lands you can get a lot of readability out of using small objects.

```ruby

PositionWithIndex = Struct.new(:value, :index)

def positions_with_index
  positions.map.with_index { |value, index| PositionWithIndex.new(value, index) }
end
```

And now elsewhere I can do stuff like:

```ruby
# Totally making this arbitrary example up on the spot.
#
positions_with_index.reduce { |empty_positions_with_indices, position_with_index|
  if position_with_index.value.nil?
    empty_positions_with_indices + [position_with_index]
  else
    empty_positions_with_indices
  end
}
```

Practically speaking, it's a simple way to get a label on structures without building out entire classes around them, in my experience it's led to much more straightforward refactors.

serpient replied:
I haven't seen this before, but very cool! Much easier to label the data :D

esmevane commented on lib/board.rb:
```diff
@@ -18,18 +18,30 @@ def full?
+        corners[1] = size - 1
+        corners[2] = corners[1] * size
+        corners[3] = corners[2] + corners[1]
+        corners
```
Does this build out a 4-member array with the index of the 4 corners?

esmevane commented on lib/player/smart_computer.rb:
```diff
@@ -0,0 +1,87 @@
+                has_potential_win = row_token_count[token_to_find] >= target_token_count
+                has_empty_position = row_token_count[Token::EMPTY] >= 1
+    
+                return row if (has_potential_win && has_empty_position)
```
First time I've seen it done like this, definitely felt like I got some brain wrinkles sorting out what was going on 🤘

esmevane commented on spec/board_spec.rb:
```diff
@@ -74,11 +74,28 @@
+
+        it 'returns a list of corners per 5x5 board size' do
+            board = Board.new(size: 5)
+            expect(board.corners).to eq [0,4,20,24]
```
Oh well, here's a spec proving what the `corners` method does :)  Cool!

esmevane commented on spec/player_spec.rb:
```diff
@@ -44,4 +53,180 @@
+            board.update(position: nil, token: nil, all_positions: smart_computer_board)
+            expect(computer_player.move(board: board, presenter: presenter)).to eq(1).or eq(2)
+        end
+    end
```
Hey, cool, testing 4x4!  How long do these tests take to run?  Just out of curiosity

serpient replied:
It says the following -- just curious, is this fast or slow for a simple game like this?
```
Finished in 0.01862 seconds (files took 0.14409 seconds to load)
88 examples, 0 failures
```

esmevane replied:
I'd say that's excellent!  💯

esmevane reviewed (approved):
👏  Really clear read, creative problem solving, and obvious understanding of some deeper functional concepts.  Great stuff!

serpient committed:
refactor per feedback

- rename play input to position to make it clearer what type of data the IO should receive
- rename message methods to remove redundant "msg" in name
- use struct to label data of positions with index
- adjust smart computer function naming. pass in parameters to each function to show what it needs to work.
